rust/src/jsonbuilder.rs

```diff
@@ -503,61 +503,34 @@ impl JsonBuilder {
     }
 
     /// Encode a string into the buffer, escaping as needed.
-    ///
-    /// The string is encoded into an intermediate vector as its faster
-    /// than building onto the buffer.
     #[inline(always)]
     fn encode_string(&mut self, val: &str) -> Result<(), JsonError> {
-        let mut buf = vec![0; val.len() * 2 + 2];
-        let mut offset = 0;
+        // reserve memory to avoid too many reallocations
+        self.buf.reserve(val.len() * 2 + 2);
         let bytes = val.as_bytes();
-        buf[offset] = b'"';
-        offset += 1;
+        self.buf.push('"');
         for &x in bytes.iter() {
-            if offset + 7 >= buf.capacity() {
-                let mut extend = vec![0; buf.capacity()];
-                buf.append(&mut extend);
-            }
             let escape = ESCAPED[x as usize];
             if escape == 0 {
-                buf[offset] = x;
-                offset += 1;
+                self.buf.push(x as char);
             } else if escape == b'u' {
-                buf[offset] = b'\\';
-                offset += 1;
-                buf[offset] = b'u';
-                offset += 1;
-                buf[offset] = b'0';
-                offset += 1;
-                buf[offset] = b'0';
-                offset += 1;
-                buf[offset] = HEX[(x >> 4 & 0xf) as usize];
-                offset += 1;
-                buf[offset] = HEX[(x & 0xf) as usize];
-                offset += 1;
+                let mut buf = vec![b'\\', b'u', b'0', b'0', 0, 0];
```

Comment:
This could be `let but = &mut &[b'\\', ...]` for maybe even tinier bit more performance..

Reply:
Then I get the compile error :
```
error[E0594]: cannot assign to `buf[_]` which is behind a `&` reference
   --> src/jsonbuilder.rs:518:17
    |
518 |                 buf[4] = HEX[(x >> 4 & 0xf) as usize];
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot assign
```

Reply:
Oops, that should be:
```
let buf = &mut [b'\\', ...];
```

Reply:
Thanks Jason